self_test/generic/complex_matrices_test/dense_complex_matrix_test.cc

```diff
@@ -0,0 +1,133 @@
+// LIC// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// LIC// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+// LIC// Lesser General Public License for more details.
+// LIC//
+// LIC// You should have received a copy of the GNU Lesser General Public
+// LIC// License along with this library; if not, write to the Free Software
+// LIC// Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
+// LIC// 02110-1301  USA.
+// LIC//
+// LIC// The authors may be contacted at oomph-lib@maths.man.ac.uk.
+// LIC//
+// LIC//====================================================================
+
+// Oomph-lib includes
+#include "generic.h"
+
+using namespace std;
+using namespace oomph;
+
+void print_complex_vector(Vector<complex<double>>& x)
+{
+  unsigned vector_length = x.size();
+  for (unsigned i = 0; i < vector_length; i++)
+  {
+    cout << x[i] << endl;
+  }
+}
+
+void print_dense_complex_matrix(DenseComplexMatrix& matrix)
+{
+  unsigned n_row = matrix.nrow();
+  unsigned n_col = matrix.ncol();
+  for (unsigned i = 0; i < n_row; i++)
+  {
+    for (unsigned j = 0; j < n_col; j++)
+    {
+      cout << matrix(i, j) << ", ";
+    }
+    cout << endl;
+  }
+}
+
+int main()
+{
+  // test default constructor
+  DenseComplexMatrix matrix_default;
+  cout << matrix_default.nrow() << endl;
+  cout << matrix_default.ncol() << endl;
+
+  constexpr unsigned long n_row = 2;
+
+  // test square matrix constructor
+  DenseComplexMatrix matrix_square(n_row);
+  cout << matrix_square.nrow() << endl;
+  cout << matrix_square.ncol() << endl;
+
+  constexpr unsigned long n_col = 3;
+
+  // test rectangular matrix constructor
+  DenseComplexMatrix matrix_rect(n_row, n_col);
+  cout << matrix_rect.nrow() << endl;
+  cout << matrix_rect.ncol() << endl;
+
+  Vector<complex<double>> values(4);
+  values[0] = complex<double>(1.0, 1.0);
+  values[1] = complex<double>(0.0, 1.0);
+  values[2] = complex<double>(2.0, 1.0);
+  values[3] = complex<double>(1.0, 1.0);
+
+  complex<double> inital_value(0.0, 1.0);
+
+  // test constructor with every entry equal to a single initial value
+  DenseComplexMatrix matrix(n_row, n_col, inital_value);
+  cout << matrix.nrow() << endl;
+  cout << matrix.ncol() << endl;
+  print_dense_complex_matrix(matrix);
+
+  // test round bracket access overloaded operator
+  matrix_square(0, 0) = values[0];
+  matrix_square(0, 1) = values[1];
+  matrix_square(1, 0) = values[2];
+  matrix_square(1, 1) = values[3];
+  print_dense_complex_matrix(matrix_square);
+
+  // test LU decomposition
+  cout << matrix_square.ludecompose() << endl;
+
+  constexpr unsigned long vector_length = n_row;
+
+  Vector<std::complex<double>> rhs(vector_length);
+  rhs[0] = complex<double>(1.0, 0.0);
+  rhs[1] = complex<double>(1.4, 1.0);
+
+  // test LU back substitution
+  // note: overwrites rhs
+  matrix_square.lubksub(rhs);
+  print_complex_vector(rhs);
+
+  Vector<std::complex<double>> x(2);
```

Comment:
Is the argument meant to be `vector_length`?

Reply:
Good spot!